Issues/4094 reader menu sites

aerych opened the pull request:
Refs #4094
This is pr number 3 of the menu refresh.  This pr introduces following management for sites to the new menu. It includes the following: 
- Ports the old following vc to swift, *mostly* intact.
- Switches from the ReaderSite entity to the ReaderSiteTopic entity for core data concerns. The ReaderSite entity should be considered deprecated from this point forward and will be removed via a clean up PR following this feature going live. 
- Adds a new field to ReaderSiteTopic to fix an issue unfollowing feeds. Fixes #5422. 
- Adds the ability to preview / browse posts from sites you follow from the new following controller. 

To test:
- On the reader menu tap the "manage" button to launch the site management
- Try following sites and feeds by typing their urls in the search bar. 
- Test the swipe to delete gesture and unfollowing sites and feeds.
- Test browsing / previewing posts from the sites/feeds you follow
- Try unfollowing a site/feed you're previewing and confirm that it does not show up when you return to the list.
- Confirm the layout is correct on the iPad.

Needs review: @frosty would you take a peek at this one?

frosty commented on WordPress/Classes/Services/ReaderTopicService.h:
```diff
@@ -27,6 +27,15 @@ extern NSString * const ReaderTopicFreshlyPressedPathCommponent;
 - (void)fetchReaderMenuWithSuccess:(void (^)())success failure:(void (^)(NSError *error))failure;
 
 /**
+ Get a list ReaderSiteTopics of the sites the user follows.
```
Super nitpicky – "get a list _of_ ReaderSiteTopics"?

frosty commented on WordPress/Classes/ViewRelated/Reader/ManageCellAccessoryView.swift:
```diff
@@ -0,0 +1,44 @@
+    var onManageTapped: ManageCellAccessoryCallback?
+
+
+    class func creaateFromNib() -> ManageCellAccessoryView {
```
`creaate`

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+    @IBOutlet var searchBar: UISearchBar!
+
+    private var refreshControl: UIRefreshControl!
+    private var isSyncing: Bool = false
```
You can ditch the type annotation here if you're initializing with a value.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+
+
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
+        tableViewController = segue.destinationViewController as? UITableViewController
```
You may want to set an identifier on the segue and check it here, for future proofing?

aerych replied:
Maybe wait and cross that bridge if necessary.  I segues have tended to be the exception rather than the rule. 
Ideally we'll eventually get rid of the child table view controller and just use a UITableView.  The only reason we have the child controller is because of the wonky behavior of the UIRefreshControl when used in other contexts, but apple has given us new API to fix that issue. :)

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+extension ReaderFollowedSitesViewController : WPTableViewHandlerDelegate
+{
+
+    func managedObjectContext() -> NSManagedObjectContext! {
```
This might be nicer as a computed property instead of a method, and return a non-optional context:

```
var managedObjectContext: NSManagedObjectContext {
    return ContextManager.sharedInstance().mainContext
}
```

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+            self?.refreshControl.endRefreshing()
+            self?.isSyncing = false
+
+            }) { [weak self] (error) in
```
I think with success and failure callbacks, it's probably better not to use trailing closure syntax, so that both closures are named? (and you could drop the brackets around `error`)

```
service.fetchFollowedSitesWithSuccess({ [weak self] in
   ...
}, failure: { [weak self] error in
  ...
})
```

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+            self?.syncSites()
+            self?.refreshFollowedPosts()
+
+            }) { (error) in
```
See note above about trailing closure syntax.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+            SVProgressHUD.showSuccessWithStatus(success)
+            self?.syncSites()
+
+        }) { (error) in
```
See note above about trailing closure syntax.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+            let buttonTitle = NSLocalizedString("OK", comment: "Button title. Acknowledges a prompt.")
+            let alert = UIAlertController(title: title, message: description, preferredStyle: .Alert)
+            alert.addCancelActionWithTitle(buttonTitle)
+            alert.presentFromRootViewController()
```
There are three blocks of code in a row here that perform almost-identical steps to display an alert with an OK button. Perhaps extract a private method like `func presentErrorAlertWithTitle(title: String, description: String)`? Then each call site could just become:

```
let title = NSLocalizedString("Could not Follow Site", comment: "Title of a prompt.")
self.showErrorAlertWithTitle(title, description: error.localizedDescription)
```

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+    }
+
+
+    func urlFromString(str: String) -> NSURL? {
```
This could go in an extension on `String`?

aerych replied:
I don't think its a good fit for that.  It makes some assumptions about what qualifies as a valid URL in this context. For instance, a host is required but strictly speaking having just a path is a valid URL (assuming a relative context).

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,332 @@
+    }
+
+
+    func fetchRequest() -> NSFetchRequest! {
```
This could also be a computed property?

frosty commented:
Hey Eric! Some small comments in the code for you. 

I also did a bunch of testing of the feature, and I've found a couple of issues. I'm not sure if any of them are pre-existing, but I'll list them all here and we can always move elsewhere if necessary.

* The search bar / text entry field at the top of the screen still says "Enter URL or tag", but you can now only enter URLs.
* It'd be nice if we showed progress / activity once you've entered a URL. When you follow / unfollow a site from the site / post view, we show an activity indicator in a HUD. Perhaps we should do this here too?
* When viewing a post and choosing Visit Site from the menu, I feel like we should take the user to the post not the site?
* Navigate into Manage > Pick a Site > Pick a post > Tap the Menu and toggle follow / unfollow, then navigate back to the site. The following status in the header doesn't update to match the new status.
* Probably related, I got the following / follow button in the Site header view very confused at one point. I could recreate this by selecting a site from the management list, unfollowing it from the button in the header, viewing a post, tapping Follow in the menu, then navigating back and hitting the button in the header again. GIF:

![follow-unfollow](https://cloud.githubusercontent.com/assets/4780/16420446/4222ef74-3d49-11e6-9fd0-24697c4f8195.gif)

Here's a GIF of the whole process:

![follow-unfollow-2](https://cloud.githubusercontent.com/assets/4780/16420454/474a09d8-3d49-11e6-95dc-7f218021cae3.gif)

* As shown in the GIF above, the Follow button in the post preview action menu is often incorrect. The simplest way I could reproduce this was: Manage sites > Add a new site > Tap into the site to preview > Tap into a Post > Open the menu. It usually says "Follow Site" for me, which is incorrect as I'm already following it.
* The Followed Sites post list doesn't automatically update when you add / remove a site. To reproduce: View Followed Sites > Head back out to Manage > Add or remove a site > Go back into Followed Sites > The list is still the same. You have to pull to refresh to update. The same happens if you unfollow a site directly from a post's menu in the Followed Sites list, or after navigating into a post.
* I feel like on iPad, the 'enter a URL' field could do with some extra padding at the top, to match the padding when you tap into a site's details.

I think that's everything! Let me know if you need any more information on anything!

aerych commented:
Hi @frosty. Thanks for the excellent review.  Code changes are made and / or replied to. 

>    The search bar / text entry field at the top of the screen still says "Enter URL or tag", but you can now only enter URLs.

Good catch. Fixed now.

>    It'd be nice if we showed progress / activity once you've entered a URL. When you follow / unfollow a site from the site / post view, we show an activity indicator in a HUD. Perhaps we should do this here too?

I like showing progress but I'm not a fan of the HUD as it blocks the UI.  Let's engage a design eye and decide on a consistent approach. I've opened #5621 for us to look into it. 

>    When viewing a post and choosing Visit Site from the menu, I feel like we should take the user to the post not the site?

Agreed. We'll look at that in #5620. 

>    Navigate into Manage > Pick a Site > Pick a post > Tap the Menu and toggle follow / unfollow, [...].
    Probably related, I got the following / follow button in the Site header view very confused at one point. [...]
    As shown in the GIF above, the Follow button in the post preview action menu is often incorrect. [...]

Following/unfollowing is a bit brittle and, probably needs a larger rethink that is out of scope for this pr.  However, I've made a couple of changes that should improve the experience.  Post stream headers should update to show the correct following status and unfollowing from the detail should now be reflected in the menu. 

>    The Followed Sites post list doesn't automatically update when you add / remove a site. [...].

I've added some new code to refresh the following topic after following or unfollowing a site. I'm not 100% sure this is the best way but lets give it a try and see. 

>    I feel like on iPad, the 'enter a URL' field could do with some extra padding at the top, to match the padding when you tap into a site's details.

Possibly.  On the other hand it matches the search bar placement and behavior in the old menu and in the new search feature.  I'm keen to leave it as is for now but let's get @jancavan's thoughts.

Ready for round two!!

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -87,7 +87,7 @@ class ReaderFollowedSitesViewController: UIViewController
 
     func configureSearchBar() {
-        let placeholderText = NSLocalizedString("Enter a URL or a tag to follow", comment: "Placeholder text prompting the user to type the name of the URL or tag they would like to follow.")
+        let placeholderText = NSLocalizedString("Enter the URL of a site to follow", comment: "Placeholder text prompting the user to type the name of the URL or tag they would like to follow.")
```
Sorry, the comment probably needs updating here too!

frosty commented:
Changes look good, @aerych!

> I like showing progress but I'm not a fan of the HUD as it blocks the UI. Let's engage a design eye and decide on a consistent approach. I've opened #5621 for us to look into it.

I definitely agree with this. Apple aren't a big fan of modal HUDs either, so if we can avoid that but show progress that'd be great :)

Happy to leave the search bar positioning for now, too.

The refreshing of the following topic has definitely helped the Followed Sites stream stay up to date after changing the following status of a site, and the follow / unfollow item in a post / site's menu always seems to be correct now. Great stuff! 👍 

However, I'm found one more bug – it doesn't seem to happen every time, but I've recreated it twice in about 5 minutes:

* Viewed Followed Sites
* Tapped into a post
* Using the menu, I unfollowed the site
* The site's posts have now (correctly) disappeared from Followed Sites
* If I go into Manage, the sites are still there. 
* In the response from `read/following/mine?meta=site,feed`, my `subscriptions` array is empty.
* If I tap into one of the sites in Manage, it still says Following.
* If I then tap the Following button in the header for that site, I get a bad access:

```
* thread #1: tid = 0x3f4eb5, 0x00000001127c6141 UIKit`__46-[UITableView _updateWithItems:updateSupport:]_block_invoke + 94, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xfffffffffffffff8)
    frame #0: 0x00000001127c6141 UIKit`__46-[UITableView _updateWithItems:updateSupport:]_block_invoke + 94
    frame #1: 0x00000001127c62a7 UIKit`__46-[UITableView _updateWithItems:updateSupport:]_block_invoke1013 + 241
    frame #2: 0x00000001127c59da UIKit`-[UITableView _updateWithItems:updateSupport:] + 2944
    frame #3: 0x00000001127be5a0 UIKit`-[UITableView _endCellAnimationsWithContext:] + 15360
  * frame #4: 0x000000010ee5518b WordPress`-[WPTableViewHandler controllerDidChangeContent:](self=0x00007fbc3c5e31f0, _cmd="controllerDidChangeContent:", controller=0x00007fbc3fa73ca0) + 91 at WPTableViewHandler.m:581
    frame #5: 0x00000001151653b6 CoreData`__77-[NSFetchedResultsController(PrivateMethods) _managedObjectContextDidChange:]_block_invoke + 4630
```

* If I restart the app, the site has gone from the Manage list.

If you need any help recreating it, let me know!

One final thing: If you unfollow all sites from the site / post detail, then when you navigate back to the Followed Sites stream it looks a little odd:

![simulator screen shot 29 jun 2016 11 18 12](https://cloud.githubusercontent.com/assets/4780/16448880/ca471770-3deb-11e6-993e-528e23d62652.png)

We should probably show the default no results view if we can in this situation.

Back to you @aerych!

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -0,0 +1,334 @@
+
+
+    func configureSearchBar() {
+        let placeholderText = NSLocalizedString("Enter the URL of a site to follow", comment: "Placeholder text prompting the user to type the name of the URL or tag they would like to follow.")
```
Sorry, this comment probably needs updating too (to remove the 'tag' reference)!

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -508,6 +510,17 @@ import WordPressComAnalytics
+    // topic's following status changed.
+    func refreshTableHeaderIfNeeded() {
+        if let siteTopic = readerTopic as? ReaderSiteTopic {
+            if let header = tableView.tableHeaderView as? ReaderStreamHeader {
```
Nitpicky, but you could combine these into a single `if`:

```
if let siteTopic = readerTopic as? ReaderSiteTopic,
  let header = tableView.tableHeaderView as? ReaderStreamHeader {
  header.configureHeader(siteTopic)
}
```

Less indentation!

aerych replied:
Will do.  I also think a `guard` would be more appropriate now that I'm giving it a second look, so I'll also make that change.

aerych commented:
Hiya @frosty 

> However, I'm found one more bug – it doesn't seem to happen every time, but I've recreated it twice in about 5 minutes

I've addressed this by checking for cached site topics and marking them unfollowed. I'm not going to worry about marking cached topics followed since the list will just be resynced when it appears and any missing sites will be filled in.  

> One final thing: If you unfollow all sites from the site / post detail, then when you navigate back to the Followed Sites stream it looks a little odd:

Nice catch.  I've added a fail-safe for this one.

Ready for another round sir!

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderFollowedSitesViewController.swift:
```diff
@@ -44,7 +44,7 @@ class ReaderFollowedSitesViewController: UIViewController
     override func viewDidLoad() {
         super.viewDidLoad()
-        self.title = NSLocalizedString("Followed Sites", comment: "Page title for the list of followed sites.")
+        self.title = NSLocalizedString("Manage Sites", comment: "Page title for the screeen to manage your list of followed sites.")
```
Tiny tiny thing – 'screeen' :)

aerych replied:
I honestly have no idea what it is with me and multiplying vowels.

frosty commented:
Changes look good! 

I'm afraid I just had another crash though, same steps. Anything I can do to help?

frosty commented:
I've tested multiple times now and I can't recreate the crash / phantom site issue! I think your Jetpack fix must've done it.

Well, I'm happy to :shipit:! Great job, @aerych!

aerych commented:
Thanks for the awesome review @frosty!
